BESTPRACTICES.md

```diff
@@ -4,17 +4,268 @@
+        }
+        ```
+
+        As you can see above w*e define type necessary fields that will be used by `handler` to interact with application state
+
+    After defining your state and messages run `make protoc` and make sure `prototool` creates your `codec.pb.go` file successfully.
+    Now we have scaffold of our application thanks to auto-generated `codec.pb.go` file.
+
+2. #### Messages
+    > [PR#2](https://github.com/iov-one/tutorial/pull/2): _Create msgs_
+    
+    This is where we wrap our auto generated message types with useful and vital functionalities. Before getting into this section I want to remind you *validation* is **EXTREMELY**  important. First create your `msg.go` file. This is where the magic will happen :grin:
+
+    Lets bind a path to our new message
+
+    ```go
+    const (
+	    pathCreateOrderBook = "order/create_book"
+    )
+    ```
+
+    After that ensure your message is a `weave.Msg`
+
+    ```go
+    var _ weave.Msg = (*CreateOrderBookMsg)(nil)
+    ```
+
+    ```go
+    func (m CreateOrderBookMsg) Validate() error {
+	    if err := m.Metadata.Validate(); err != nil {
+		    return errors.Wrap(err, "metadata")
+	    }
+	    if err := validID(m.MarketID); err != nil {
+		    return errors.Wrap(err, "market id")
+	    }
+	    if !coin.IsCC(m.AskTicker) {
+		    return errors.Wrapf(errors.ErrCurrency, "Invalid Ask Ticker: %s", m.AskTicker)
+	    }
+	    if !coin.IsCC(m.BidTicker) {
+		    return errors.Wrapf(errors.ErrCurrency, "Invalid Bid Ticker: %s", m.BidTicker)
+	    }
+	    if m.BidTicker <= m.AskTicker {
+		    return errors.Wrapf(errors.ErrCurrency, "ask (%s) must be before bid (%s)", m.AskTicker, m.BidTicker)
+	    }
+	    return nil
+    }
+    ```
+
+    ```go
+    func validID(id []byte) error {
+	    if len(id) == 0 {
+		    return errors.Wrap(errors.ErrEmpty, "id missing")
+	    }
+	    if len(id) != 8 {
+		    return errors.Wrap(errors.ErrInput, "id is invalid length (expect 8 bytes)")
+	    }
+	    return nil
+    }
+    ```
+
+    You must have notice we even validate if ```ID```'s lenght is not 0 and equal to 8 and tickers are actually string tickers. **Remember** more validation more solid your application is.
+
+3. #### Models
+    > [PR#6](https://github.com/iov-one/tutorial/pull/6): _Create models_
+
+    We defined our state in [codec section](#codec). In order to use models in weave we have to wrap our model with some functionalities and enforce it is a **morm.Model**
+
+    Ensure our *OrderBook* fulfills **morm.Model**
+
+    ```go
+    var _ morm.Model = (*OrderBook)(nil)
+    ```
+    
+    Define **SetID** function
+
+    ```go
+    func (o *OrderBook) SetID(id []byte) error {
+	    o.ID = id
+	    return nil
+    }
+    ```
+
+    In order to our model to fulfill **Model** it must be [Clonable](https://github.com/iov-one/weave/blob/master/orm/interfaces.go#L34).
+    
+    This is how you ensure it:
+
+    ```go
+    func (o *OrderBook) Copy() orm.CloneableData {
```

Comment:
For future work: can we use `reflect.DeepCopy`? I think this would make it easier to understand.

Reply:
Could not understand the concept of the comment. You mean for future work as learn how `reflect.DeepCopy` works and explain why to use or not, Or investigate in future if we can use `reflect.DeepCopy` code here?

Reply:
>  Or investigate in future if we can use reflect.DeepCopy code here?

That one. I think it will do the same, but be a bit slower. It would be good to validate it is functionally equivalent to hand-drawn code. Then look at the benchmarks. If it is not much overhead, we should propose using it by default.

But that is another story... just needed a place to write down the thoughts. Nothing to impact your doc work at all